### PR TITLE
New gui feature

### PR DESCRIPTION
**Pull request by michael4180**
I set up a new file to work on adding code for a gui.  I installed kivy to do this.  I tried using tkinter, which is part of the standard library, but i kept getting an error - module not found.  In reviewing this, the workaround i found was to uninstall and reinstall python, but when i would do this, i would get dependency warnings.  So I didn't uninstall.  I instead went with kivy and installed it.  This worked fine but now when I run my script for running a gui, I'm getting a blank box with no data showing up.  Not sure what to do to fix this.  Let me know your thoughts..thx

### CODE REVIEW

File: gui_weather_app.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        return self.layout
+
+    def display_weather(self, instance):
+        city_name = self.city_name_input.text
+        weather_data = get_weather(city_name)
+        if weather_data and weather_data.get("cod") == 200:
+            forecast_data = get_weather(city_name, forecast=True)
+            weather_text = f"Weather in {weather_data['name']}: {weather_data['weather'][0]['description']}\n"
+            weather_text += f"Temperature: {weather_data['main']['temp']}°C, Humidity: {weather_data['main']['humidity']}%\n"
+            weather_text += "\n5-Day Weather Forecast:\n"
```
Comment (bbelderbos) [inline gui_weather_app.py:47]: you can also use `"\n".join(list_of_strings)` here

Comment (bbelderbos) [reply gui_weather_app.py:47]: = more elegant, but also faster if you do this concatenation a lot, see https://pybit.es/articles/faster-python/ - 2.

**Comment by bbelderbos** [comment]
Does work for me?

image

Needs some styling though, both input field as output displayed. 

Great work!

Can you also look at the other issues and log more if needed?

Talk tomorrow ...

**Review by bbelderbos — APPROVED** [review]

File: gui_weather_app.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
+from kivy.app import App
+from kivy.uix.boxlayout import BoxLayout
+from kivy.uix.label import Label
+from kivy.uix.textinput import TextInput
+from kivy.uix.button import Button
+import requests
+
+from config import openweather_api_key as api_key
```
[the remark this answers is not in the document]
Comment (michael4180) [reply gui_weather_app.py:8]: thx bob.  Let's review in the morning

File: gui_weather_app.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+
+def get_weather(city_name, forecast=False):
```
[the remark this answers is not in the document]
Comment (michael4180) [reply gui_weather_app.py:11]: ok, thx

File: gui_weather_app.py (hunk 1 above)
Comment (michael4180) [reply gui_weather_app.py:47]: got it..thx

File: weather_app.py [hunk 4]
```diff
@@ -21,10 +24,26 @@ def display_weather(weather_data):
...
+        time = item['dt_txt']
+        temp = item['main']['temp']
+        description = item['weather'][0]['description']
+        print(f"{time}: {description}, Temp: {temp}°C")
+
 def main():
-    city_name = input("Enter a city name: ")
-    weather_data = get_weather(city_name)
-    display_weather(weather_data)
+    while True:
```
[the remark this answers is not in the document]
Comment (michael4180) [reply weather_app.py:36]: thx

**Comment by michael4180** [comment]
thx bob..not able to code today but did review code earlier for about an hour in a book that provides small projects to work on.  The programs are scripted and you can just review and play with the programs in your ide.  Started doing this.  Seems useful.  The book is "The Big Book of Small Python Projects".   Speak to you tomrrow..thx

**Comment by bbelderbos** [comment]
> thx bob..not able to code today but did review code earlier for about an hour in a book that provides small projects to work on. The programs are scripted and you can just review and play with the programs in your ide. Started doing this. Seems useful. The book is "The Big Book of Small Python Projects". Speak to you tomrrow..thx

Sounds useful yes. Is this the corresponding github repo? https://github.com/asweigart/the-big-book-of-small-python-projects - if so, let me know which modules you're looking at. 

I think it will also be beneficial to, apart from this project here, code the recommended bite paths here: https://github.com/pybites/PDI_MikeG/issues/7#issuecomment-2056956958 - you can document any questions or learnings in that repo + issue.

Talk in a bit ...

File: flask_weather_app.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+@app.route('/', methods=['GET', 'POST'])
+def index():
+    weather = None
+    forecast = None
+    if request.method == 'POST':
+        city_name = request.form['city']
+        weather = get_weather(city_name)
+        if weather:
+            forecast = get_weather(city_name, forecast=True)
```
Comment (bbelderbos) [inline flask_weather_app.py:29]: nice re-use of your functions

**Review by bbelderbos — CHANGES_REQUESTED** [review]
This works great, some minor changes before we merge and move on, discuss in a bit ...

File: flask_weather_app.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
+from flask import Flask, render_template, request
+import requests
+
+app = Flask(__name__)
```
[the remark this answers is not in the document]
Comment (bbelderbos) [reply flask_weather_app.py:4]: done, commited on main

File: flask_weather_app.py (hunk 5 above)
Comment (bbelderbos) [reply flask_weather_app.py:29]: already having an issue for this: #13 - can discuss there ...

File: gui_weather_app.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (bbelderbos) [reply gui_weather_app.py:8]: logged in an issue already, please implement as one of the next steps ...
#6

File: gui_weather_app.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (bbelderbos) [reply gui_weather_app.py:11]: opened issue: #19

File: templates/weather.html [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+        <button type="submit">Check Weather</button>
+    </form>
+    {% if weather %}
+        <h2>Current Weather in {{ weather.name }}</h2>
+        <p>{{ weather.weather[0].description }}</p>
+        <p>Temperature: {{ weather.main.temp }}°C</p>
+        <p>Humidity: {{ weather.main.humidity }}%</p>
+        {% if forecast %}
+            <h3>5-Day Forecast:</h3>
+            {% for item in forecast.list %}
```
[the remark this answers is not in the document]
Comment (bbelderbos) [reply templates/weather.html:21]: logged: #17

File: flask_weather_app.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+# Assuming you have your openweather_api_key in a config module
+from config import openweather_api_key as api_key
+
+def get_weather(city_name, forecast=False):
```
[the remark this answers is not in the document]
Comment (bbelderbos) [reply flask_weather_app.py:9]: logged: #19

File: flask_weather_app.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    base_url = "http://api.openweathermap.org/data/2.5/weather" if not forecast else "http://api.openweathermap.org/data/2.5/forecast"
```
Comment (bbelderbos) [inline flask_weather_app.py:10]: use multiline if/else (no ternary) and abstract repeated base url in constant at top of the module.

**Pull request merged**